### PR TITLE
Game mechanism

### PR DESCRIPTION
**Pull request by Chroq**
## Description

In this PR, we want to be able to initialize a new online game.

Trello: https://trello.com/c/3W8vHuQw/4-2-as-norbert-i-want-to-play-online-with-philippe

- [X] Create an online game
  - [X] Create an API route
    -  [X] Post request to create a new game
- [X] Get back a game to load the board
- [X] Create an API route
  - [X] Get request to get back a game by ID
- [X] Adapt existing modal to create a game

NB : There's a big refactoring in this pull request. The "Playboard" component had too much logic and I had to split it into several files. 

image
_new menu_

image
_http://localhost:3000/board/online?id=e14e6fac-f7d0-4761-9d63-9544ba0f254e_

### CODE REVIEW

File: components/boards/Playboard.js [hunk 1]
```diff
@@ -1,36 +1,100 @@
...
+  sizeParameter,
+  idParameter,
+  player1NicknameParameter,
+  player2NicknameParameter
+) => {
+  const [game, setGame] = useState();
+
+  useEffect(() => {
+    if (sizeParameter) {
+      return createNewGame(
```
Comment (djhi) [inline components/boards/Playboard.js:42]: You don't set the `game` state when this function promise resolves

File: components/boards/Playboard.js [hunk 2]
```diff
@@ -1,36 +1,100 @@
...
+        sizeParameter,
+        onlineParameter,
+        player1NicknameParameter
+      );
+    } else if (idParameter) {
+      return loadExistingGame(
```
Comment (djhi) [inline components/boards/Playboard.js:48]: Same

File: components/boards/Playboard.js (hunk 1 above)
Comment (djhi) [reply components/boards/Playboard.js]: My mistake, `createNewGame` does not return a promise but you should still use its result to set the game

File: components/boards/Playboard.js (hunk 2 above)
Comment (djhi) [reply components/boards/Playboard.js]: This one returns a promise however

File: components/boards/Playboard.js [hunk 3]
```diff
@@ -1,36 +1,100 @@
...
+      );
+    } else if (idParameter) {
+      loadExistingGame(
+        idParameter,
+        player2NicknameParameter,
+        onlineParameter
+      ).then(function (game) {
+        setGame(game);
+      });
+    }
```
Comment (djhi) [inline components/boards/Playboard.js:53]: What happens if you neither `sizeParameter` nor `idParameter` ?

**Comment by fzaninotto** [comment]
Can you please add a checklist to PRs in WIP, as well as screenshots of your progress so far?

**Comment by Chroq** [comment]
I have updated the description. I didn't expect to have all this changes, sorry about that.

File: components/boards/Playboard.js [hunk 4]
```diff
@@ -1,48 +1,58 @@
...
+  return {
+    grid: game.grid,
+    size: Math.sqrt(game.grid.length),
+    player: game.player,
+    winner: game.winner,
+  };
+}
+
+function Playboard({ game, ...props }) {
+  const [{ grid, player, winner, size }, dispatch] = useReducer(
```
Comment (djhi) [inline components/boards/Playboard.js:41]: Careful there, this reducer has many side effects. We discussed this and you have to refactor this in another PR. Ideally, the Playboard component should only be responsible to display the game and handle UI interactions. Its parents are responsible for applying the effects of thoses interractions, save the game either locally or remotely and resend through props the updated game.

File: pages/api/games/[uuid].js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      return patch(req, res);
+  }
+};
+
+function get(req, res) {
+  const {
+    query: { uuid },
+  } = req;
+
+  getGameRepository()
```
Comment (fzaninotto) [inline pages/api/games/[uuid].js:22]: How can it work if you don't return the Promise?

Comment (fzaninotto) [reply pages/api/games/[uuid].js:22]: Nevermind, you don't need to `return` in next. But you don't need it line 34 either

File: .eslintrc [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      "parser": "babel-eslint"
+    },
+    "env": {
+      "browser": true,
+      "es6": true
+    },
+    "rules": {
+        "eqeqeq": "warn",
+        "indent": ["warn", 2, { "SwitchCase": 1 }],
+        "multiline-ternary": ["warn", "always-multiline"]
```
[the remark this answers is not in the document]
Comment (Chroq) [reply .eslintrc:15]: Noted, thanks for the link

File: Makefile [hunk 7]
```diff
@@ -7,7 +7,7 @@ help:
 
 install: ## Install project's dependencies
 	@echo "Install project deps"
-	yarn install
+	yarn install --frozen-lockfile
```
[the remark this answers is not in the document]
Comment (Chroq) [reply Makefile:10]: Noted, I'll add another make command for CI

File: components/boards/Playboard.js [hunk 8]
```diff
@@ -1,48 +1,58 @@
...
+  calculateTopPosition,
+  getBoardRatio,
+  getHexagonHeight,
+  getHexagonWidth,
+} from "./position.js";
 
 const ERROR_NOT_FOUND_GAME = `Can't find the game with ID`;
+const DEFAULT_SIZE = 11;
+
+function init({ game }) {
```
[the remark this answers is not in the document]
Comment (Chroq) [reply components/boards/Playboard.js:30]: Moved in the new PR, thanks

File: components/boards/Playboard.js (hunk 4 above)
Comment (Chroq) [reply components/boards/Playboard.js:41]: Noted with thanks, I'm making corrections in a new PR.

File: pages/api/games/[uuid].js (hunk 5 above)
Comment (Chroq) [reply pages/api/games/[uuid].js:22]: Indeed, I'll remove this useless `return`, thanks

**Pull request merged**